### DISCUSSION TITLE
[General] How to get last_verified_block ?

### DISCUSSION THREAD

**Discussion by amedrontadora** [zkSync-Community-Hub/zksync-developers#14] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

Is there a way to obtain the latest validated l1 batch height?
I can only find the latest batch number by "zks_L1BatchNumber", but I cannot locate the most recently validated block height.

### What is the specific issue or error you're encountering?

Unable to directly obtain the latest provn block height through code

### Can you share the error messages or logs you're receiving, if any?

.

### Have you made any recent changes to the contract before encountering this issue?

.

### Are there any external libraries or contracts that your contract interacts with?

.

### Can you provide the relevant portions of your contract code where the issue is occurring?

.

### Have you tried to isolate the problem, and if so, what were the results?

.

### What steps have you already taken to try to resolve the issue?

I have tried to access the zkSync Era: Validator Timelock contract or filter proof transactions in L1, but the contract did not emit any events.

**Comment by uF4No** [comment, marked as answer]
For L2 blocks, you can use the `getBlock` method from the JS SDK and pass the "finalized" block tag. Example:

```js
const l2provider = new Provider("https://zksync2-testnet.zksync.dev");
console.log(`${await l2provider.getBlockNumber()}`);

const latesConfirmed = await l2provider.getBlock("finalized");
console.log(latesConfirmed.number);
```

For batches, you'd need to call any of the following L1 contract methods from the [GettersFacet](https://etherscan.io/address/0xF1fB730b7f8E8391B27B91f8f791e10E4a53CEcc#readContract):

```
function getTotalBlocksCommitted() external view returns (uint256);

function getTotalBlocksVerified() external view returns (uint256);

function getTotalBlocksExecuted() external view returns (uint256);
```
You can obtain [the ABI of the GettersFacet from here](https://etherscan.io/address/0xF1fB730b7f8E8391B27B91f8f791e10E4a53CEcc#code), and then use a script like following:

```js
 const contract = new Contract(
    "0x32400084c286cf3e17e7b677ea9583e60a000324", // zkSync Era mainnet proxy
    GETTERSFACET,
    l1provider
  );

  console.log(`Verified ${await contract.getTotalBlocksVerified()}`);
```
just keep in mind you'd need to call the `GettersFacet` via the proxy contract. 

Both addresses for mainnet and testnet are:

**Mainnet**
- DiamondProxy: [0x32400084c286cf3e17e7b677ea9583e60a000324](https://etherscan.io/address/0x32400084c286cf3e17e7b677ea9583e60a000324#code)
- GettersFacet: [0xF1fB730b7f8E8391B27B91f8f791e10E4a53CEcc](https://etherscan.io/address/0xF1fB730b7f8E8391B27B91f8f791e10E4a53CEcc#code)

**Testnet**
- DiamondProxy: [0x1908e2BF4a88F91E4eF0DC72f02b8Ea36BEa2319](https://goerli.etherscan.io/address/0x1908e2BF4a88F91E4eF0DC72f02b8Ea36BEa2319#code)
- GettersFacet: [0x71b3Ffda716Ef5df529FA89a8BBb8D16676fD47f](https://goerli.etherscan.io/address/0x71b3Ffda716Ef5df529FA89a8BBb8D16676fD47f#code)

**Comment by newtondav** [comment]
To obtain the latest validated L1 batch height in zkSync, you can use the zks_getL1BlockRange API. This will provide you with information about the latest validated block height on the Ethereum mainnet, which corresponds to the latest finalized L1 batch. You can use this information to track the synchronization between L1 and L2.

**Comment by fullertasha** [comment]
Suggestion:
To obtain the latest validated L1 batch height, you can monitor events emitted by zkSync contracts or use zkSync APIs.

**Comment by unticcutecerin** [comment]
The user is trying to obtain the latest validated L1 batch height, but is only able to find the latest batch number through "zks_L1BatchNumber" and is unable to locate the most recently validated block height. The user has tried accessing the zkSync Era: Validator Timelock contract and filtering proof transactions in L1, but has not been successful. Without further information or access to the code repository, it is difficult to provide a specific solution. However, it may be helpful for the user to reach out to the zkSync support team or consult the zkSync documentation for guidance on how to obtain the latest validated L1 batch height.
